// react-pdf-js/lib/index.js
'use strict';

exports.__esModule = true;
exports.ReactPdfJs = undefined;

var _pdfjsDist = require('pdfjs-dist');

var _pdfjsDist2 = _interopRequireDefault(_pdfjsDist);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var usePrevious = function usePrevious(value) {
  var ref = (0, _react.useRef)();
  (0, _react.useEffect)(function () {
    ref.current = value;
  });
  return ref.current;
}; /**
    * @class ReactPdfJs
    */
var ReactPdfJs = exports.ReactPdfJs = function ReactPdfJs(_ref) {
  var file = _ref.file,
      onDocumentComplete = _ref.onDocumentComplete,
      scale = _ref.scale,
      rotate = _ref.rotate,
      page = _ref.page,
      cMapUrl = _ref.cMapUrl,
      cMapPacked = _ref.cMapPacked,
      className = _ref.className,
      workerSrc = _ref.workerSrc;

  var _useState = (0, _react.useState)(null),
      pdf = _useState[0],
      setPdf = _useState[1];

  var canvasEl = (0, _react.useRef)(null);

  // do our initial setup
  (0, _react.useEffect)(function () {
    _pdfjsDist2.default.GlobalWorkerOptions.workerSrc = workerSrc;
    var loadingTask = _pdfjsDist2.default.getDocument({ url: file, cMapUrl: cMapUrl, cMapPacked: cMapPacked });
    loadingTask.promise.then(function (document) {
      setPdf(document);
      if (onDocumentComplete) {
        onDocumentComplete(document._pdfInfo.numPages); // eslint-disable-line
      }
      document.getPage(page).then(function (p) {
        return drawPDF(p);
      });
    });
  }, []);

  // see if anything has changed
  var oldPage = usePrevious(page);
  var oldScale = usePrevious(scale);
  var oldRotate = usePrevious(rotate);
  (0, _react.useEffect)(function () {
    if (pdf && (oldPage !== page || oldScale !== scale || oldRotate !== rotate)) {
      pdf.getPage(page).then(function (p) {
        return drawPDF(p);
      });
    }
  }, [page, scale, rotate]);

  // draw a page of the pdf
  var drawPDF = function drawPDF(page) {
    // Because this page's rotation option overwrites pdf default rotation value,
    // calculating page rotation option value from pdf default and this component prop rotate.
    var rotation = rotate === 0 ? page.rotate : page.rotate + rotate;
    var viewport = page.getViewport({ scale: scale, rotation: rotation });
    var canvas = canvasEl.current;
    var canvasContext = canvas.getContext('2d');
    canvas.height = viewport.height;
    canvas.width = viewport.width;
    var renderContext = {
      canvasContext: canvasContext,
      viewport: viewport
    };
    page.render(renderContext);
  };

  return _react2.default.createElement('canvas', { ref: canvasEl, className: className });
};

ReactPdfJs.propTypes = process.env.NODE_ENV !== "production" ? {
  file: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.object]).isRequired,
  page: _propTypes2.default.number,
  onDocumentComplete: _propTypes2.default.func,
  scale: _propTypes2.default.number,
  rotate: _propTypes2.default.oneOf([0, 90, 180, 270]),
  cMapUrl: _propTypes2.default.string,
  cMapPacked: _propTypes2.default.bool,
  className: _propTypes2.default.string,
  workerSrc: _propTypes2.default.string
} : {};

ReactPdfJs.defaultProps = {
  page: 1,
  onDocumentComplete: null,
  scale: 1,
  rotate: 0,
  cMapUrl: '../node_modules/pdfjs-dist/cmaps/',
  cMapPacked: false,
  className: '',
  workerSrc: '//cdnjs.cloudflare.com/ajax/libs/pdf.js/2.1.266/pdf.worker.js'
};

exports.default = ReactPdfJs;
